File: project_euler/problem_004/solution.py
"""Project Euler Problem 4: Largest Palindrome Product"""

def is_palindrome(n: int) -> bool:
    s = str(n)
    return s == s[::-1]

def solve() -> int:
    best = 0
    # One factor must be divisible by 11 (Theorem 1 + Euclid's lemma)
    for a in range(990, 99, -11):
        for b in range(999, a - 1, -1):
            prod = a * b
            if prod <= best:
                break
            if is_palindrome(prod):
                best = prod
                break
    return best

answer = solve()
assert answer == 906609
print(answer)

File: project_euler/problem_004/solution.cpp
#include <bits/stdc++.h>
using namespace std;

bool is_palindrome(int n) {
    string s = to_string(n);
    int lo = 0, hi = (int)s.size() - 1;
    while (lo < hi) {
        if (s[lo++] != s[hi--]) return false;
    }
    return true;
}

int main() {
    int best = 0;
    // By Theorem 1 + Euclid's lemma: one factor must be divisible by 11
    for (int x = 990; x >= 110; x -= 11) {
        for (int y = 999; y >= x; y--) {
            int P = x * y;
            if (P <= best) break;
            if (is_palindrome(P)) {
                best = P;
                break;
            }
        }
    }
    cout << best << endl;
    return 0;
}
